nfs/src/rpc_task.cpp

```diff
@@ -886,13 +972,58 @@ static void createfile_callback(
         task->get_client()->get_nfs_inode_from_ino(ino);
     const int status = task->status(rpc_status, NFS_STATUS(res));
 
+    INJECT_CREATE_FH_POPULATE_FAILURE(res, task);
+
     /*
      * Now that the request has completed, we can query libnfs for the
      * dispatch time.
      */
     task->get_stats().on_rpc_complete(rpc_get_pdu(rpc), NFS_STATUS(res));
 
     if (status == 0) {
+        if (!res->CREATE3res_u.resok.obj.handle_follows) {
+            /*
+             * If the server doesn't send the filehandle (which is perfectly
+             * valid), make a LOOKUP RPC request to query the filehandle and
+             * pass that to fuse.
+             */
+            AZLogWarn("CreateFile failed to return filehandle, req={}, "
+                "parent_ino={}, name={}. Issuing lookup.",
+                fmt::ptr(task->get_fuse_req()),
+                task->rpc_api->create_task.get_parent_ino(),
+                task->rpc_api->create_task.get_file_name());
+
+            struct fuse_req *req = task->get_fuse_req();
+            fuse_ino_t parent_ino = task->rpc_api->create_task.get_parent_ino();
+            char *file_name = ::strdup(
+                task->rpc_api->create_task.get_file_name());
+            struct fuse_file_info *fusefile =
+                task->rpc_api->create_task.get_fuse_file();
+
+            /*
+             * Free the current task as the response will be sent by the
+             * lookup task made below.
+             * Note: task should not be accessed after this.
+             */
+            task->free_rpc_task();
+
+            struct rpc_task *lookup_tsk =
+                    task->get_client()->get_rpc_task_helper()->alloc_rpc_task(
+                        FUSE_LOOKUP);
+            lookup_tsk->init_lookup(
+                req,
+                file_name,
+                parent_ino);
+
+            // Mark the lookup task as being called to populate the fh by create.
+            lookup_tsk->rpc_api->lookup_task.set_called_for_optype(FUSE_CREATE);
+            lookup_tsk->rpc_api->lookup_task.set_fuse_file(fusefile);
+
+            lookup_tsk->run_lookup();
+            ::free(file_name);
+            return;
+        }
```

Comment:
This requirement of issuing another task on behalf of an original task is a general req and can be used in multiple places. Let's make it standardized.
I suggest calling this a proxy operation, so we add

init_proxy_lookup();
run_proxy_lookup();

and hide all these details inside those, and from here just call these.

You can use the common functions in struct api_task_info
then we add called_for_optype as proxy_optype in api_task_info

similarly we can have proxy getattr